runtime-interface/src/lib.rs

```diff
@@ -70,14 +74,85 @@ impl PassBy for LazyPagesProgramContext {
 
 #[derive(Debug, Clone, Encode, Decode)]
 #[codec(crate = codec)]
-pub struct LazyPagesRuntimeContext {
+pub struct LazyPagesInitContext {
     pub page_sizes: Vec<u32>,
     pub global_names: Vec<LimitedStr<'static>>,
     pub pages_storage_prefix: Vec<u8>,
 }
 
-impl PassBy for LazyPagesRuntimeContext {
-    type PassBy = Codec<LazyPagesRuntimeContext>;
+impl From<gear_lazy_pages_common::LazyPagesInitContext> for LazyPagesInitContext {
+    fn from(ctx: gear_lazy_pages_common::LazyPagesInitContext) -> Self {
+        let gear_lazy_pages_common::LazyPagesInitContext {
+            page_sizes,
+            global_names,
+            pages_storage_prefix,
+        } = ctx;
+
+        Self {
+            page_sizes,
+            global_names,
+            pages_storage_prefix,
+        }
+    }
+}
+
+impl From<LazyPagesInitContext> for gear_lazy_pages_common::LazyPagesInitContext {
+    fn from(ctx: LazyPagesInitContext) -> Self {
+        let LazyPagesInitContext {
+            page_sizes,
+            global_names,
+            pages_storage_prefix,
+        } = ctx;
+
+        Self {
+            page_sizes,
+            global_names,
+            pages_storage_prefix,
+        }
+    }
+}
+
+impl PassBy for LazyPagesInitContext {
+    type PassBy = Codec<LazyPagesInitContext>;
+}
+
+#[derive(Debug, derive_more::Display)]
+enum SpIoProgramStorageError {
+    #[display(fmt = "Page data in storage must contain {expected} bytes, actually has {actual}")]
+    InvalidPageDataSize { expected: u32, actual: u32 },
+}
+
+#[derive(Debug, Default)]
+struct SpIoProgramStorage;
+
+#[cfg(feature = "std")]
+impl LazyPagesPagesStorage for SpIoProgramStorage {
+    fn page_exists(&self, prefix: &PagePrefix, page: GearPage) -> bool {
+        let key = prefix.key_for_page(page);
+        sp_io::storage::exists(&key)
+    }
+
+    fn load_page(
+        &mut self,
+        page_sizes: &PageSizes,
```

Comment:
pages may be different sizes?

Reply:
Please remove this argument. Yes, page can change size, but in LazyPagesStorage the only one place you use it is for error checking, which is not semantically important, also you can use buffer size, supposing that whoever call  `load_page` must provide exact size as page buffer has.